sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeDestructor.scala

```diff
@@ -0,0 +1,259 @@
+        tyCon: Ref.TypeConName,
+        pkgName: Ref.PackageName,
+        cons: ArraySeq[Ref.Name],
+    ) extends TypeF[Nothing] {
+      private[this] lazy val consRankMap = cons.view.zipWithIndex.toMap
+
+      def consRank(cons: Ref.Name): Either[LookupError, Int] = {
+        def ref = Reference.DataEnumConstructor(tyCon, cons)
+
+        consRankMap.get(cons).toRight(LookupError.NotFound(ref, ref))
+      }
+    }
+  }
+
+  sealed abstract class Error extends Product with Serializable
+
+  object Error {
+    final case class LookupError(error: language.LookupError) extends Error
+
+    final case class TypeError(msg: String) extends Error
+  }
+
+  def apply(pkgInterface: PackageInterface): TypeDestructor =
+    new TypeDestructorImpl(pkgInterface)
+
+}
+
+private final class TypeDestructorImpl(pkgInterface: PackageInterface) extends TypeDestructor {
+  self =>
+
+  import TypeDestructor.TypeF
+  import TypeF._
+
+  override type Type = Ast.Type
+
+  override def wrap(typ: Ast.Type): Type = typ
+
+  private var cache: Map[Ast.Type, TypeF[Ast.Type]] = Map.empty
+
+  override def destruct(state: Type): Either[TypeDestructor.Error, TypeF[Type]] =
+    cache.get(state) match {
+      case Some(typeF) =>
+        Right(typeF)
+      case None =>
+        val result = go(state, List.empty)
+        val _ = result.foreach(typeF => cache = cache.updated(state, typeF))
+        result
+    }
+
+  private def go(
+      typ0: Ast.Type,
+      args: List[Ast.Type],
+  ): Either[TypeDestructor.Error, TypeF[Type]] = {
+    def prettyType = args.foldLeft(typ0)(Ast.TApp).pretty
+
+    def unsupportedType = TypeDestructor.Error.TypeError(s"unsupported type $prettyType")
+
+    def wrongType = TypeDestructor.Error.TypeError(s"wrong type $prettyType")
+
+    typ0 match {
+      case Ast.TSynApp(tysyn, synArgs) =>
+        for {
+          synDef <- pkgInterface.lookupTypeSyn(tysyn).left.map(TypeDestructor.Error.LookupError)
+          params = synDef.params
+          subst <- Either.cond(
+            params.length == synArgs.length,
+            params.toSeq.view.map(_._1) zip synArgs.toSeq.view,
+            TypeDestructor.Error.TypeError(s"wrong number of argument for type synonym $tysyn"),
+          )
+          typ = AstUtil.substitute(synDef.typ, subst)
+          r <- go(typ, args)
+        } yield r
+      case Ast.TTyCon(tycon) =>
+        for {
+          pkg <- pkgInterface
+            .lookupPackage(tycon.packageId)
+            .left
+            .map(TypeDestructor.Error.LookupError)
+          pkgName = pkg.metadata.name
+          dataDef <- pkgInterface.lookupDataType(tycon).left.map(TypeDestructor.Error.LookupError)
+          params = dataDef.params
+          subst <- Either.cond(
+            params.length == args.length,
+            params.toSeq.view.map(_._1) zip args.view,
+            TypeDestructor.Error.TypeError(s"wrong number of argument for $tycon"),
+          )
+          destructor <- dataDef.cons match {
```

Comment:
Minor: I would call it `destructed` or `destructedTyp0` rather than `destructor`.